pkg/buildscript/internal/raw/marshal.go

```diff
@@ -0,0 +1,101 @@
+package raw
+
+import (
+	"encoding/json"
+	"errors"
+	"strings"
+	"time"
+
+	"github.com/ActiveState/cli/internal/constants"
+	"github.com/ActiveState/cli/internal/errs"
+	"github.com/ActiveState/cli/internal/locale"
+	"github.com/ActiveState/cli/internal/rtutils/ptr"
+	"github.com/ActiveState/cli/pkg/buildscript/internal/buildexpression"
+	"github.com/alecthomas/participle/v2"
+	"github.com/go-openapi/strfmt"
+)
+
+// Marshal converts our Raw structure into a the ascript format
+func (r *Raw) Marshal() ([]byte, error) {
+	be, err := r.MarshalBuildExpression()
+	if err != nil {
+		return nil, errs.Wrap(err, "Could not marshal build expression")
+	}
+
+	expr, err := buildexpression.Unmarshal(be)
+	if err != nil {
+		return nil, errs.Wrap(err, "Could not unmarshal build expression")
+	}
+
+	return []byte(marshalFromBuildExpression(expr, r.AtTime)), nil
+}
```

Comment:
exploding-head-emojii

Reply:
Yeah this took a while to untangle, and it's still pretty confusing. I think ultimately this will be fixed by having the ascript package and by moving most of the interpretation logic to buildscripts rather than buildexpressions.